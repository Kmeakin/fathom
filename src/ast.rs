//! The syntax of our data description language

use source::Span;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Kind {
    Type,
}

/// A type definition
///
/// ```plain
/// Point = {
///     x : u16,
///     y : u16,
/// }
/// ```
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Definition {
    pub span: Span,
    pub name: String,
    pub ty: Type,
}

impl Definition {
    pub fn new<Sp, S>(span: Sp, name: S, ty: Type) -> Definition
    where
        Sp: Into<Span>,
        S: Into<String>,
    {
        let span = span.into();
        let name = name.into();

        Definition { span, name, ty }
    }
}

/// An expression
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Expr {
    /// An integer constant: `0`, `1`, `2`, ...
    Const(Span, u64),
    /// A variable, referring to an integer that exists in the current context: `len`, `num_tables`
    Var(Span, String),
    /// Integer negation: `-x`
    Neg(Span, Box<Expr>),
    /// Integer addition: `x + y`
    Add(Span, Box<Expr>, Box<Expr>),
    /// Integer subtraction: `x - y`
    Sub(Span, Box<Expr>, Box<Expr>),
    /// Integer multiplication: `x * y`
    Mul(Span, Box<Expr>, Box<Expr>),
    /// Integer division: `x / y`
    Div(Span, Box<Expr>, Box<Expr>),
}

impl Expr {
    /// An integer constant: `0`, `1`, `2`, ...
    pub fn const_<Sp>(span: Sp, value: u64) -> Expr
    where
        Sp: Into<Span>,
    {
        Expr::Const(span.into(), value)
    }

    /// A variable, referring to an integer that exists in the current context: `len`, `num_tables`
    pub fn var<Sp, S>(span: Sp, name: S) -> Expr
    where
        Sp: Into<Span>,
        S: Into<String>,
    {
        Expr::Var(span.into(), name.into())
    }

    /// Integer negation: `-x`
    pub fn neg<Sp, T>(span: Sp, x: T) -> Expr
    where
        Sp: Into<Span>,
        T: Into<Box<Expr>>,
    {
        Expr::Neg(span.into(), x.into())
    }

    /// Integer addition: `x + y`
    pub fn add<Sp, T, U>(span: Sp, x: T, y: U) -> Expr
    where
        Sp: Into<Span>,
        T: Into<Box<Expr>>,
        U: Into<Box<Expr>>,
    {
        Expr::Add(span.into(), x.into(), y.into())
    }

    /// Integer subtraction: `x - y`
    pub fn sub<Sp, T, U>(span: Sp, x: T, y: U) -> Expr
    where
        Sp: Into<Span>,
        T: Into<Box<Expr>>,
        U: Into<Box<Expr>>,
    {
        Expr::Sub(span.into(), x.into(), y.into())
    }

    /// Integer multiplication: `x * y`
    pub fn mul<Sp, T, U>(span: Sp, x: T, y: U) -> Expr
    where
        Sp: Into<Span>,
        T: Into<Box<Expr>>,
        U: Into<Box<Expr>>,
    {
        Expr::Mul(span.into(), x.into(), y.into())
    }

    /// Integer division: `x / y`
    pub fn div<Sp, T, U>(span: Sp, x: T, y: U) -> Expr
    where
        Sp: Into<Span>,
        T: Into<Box<Expr>>,
        U: Into<Box<Expr>>,
    {
        Expr::Div(span.into(), x.into(), y.into())
    }
}

#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub enum TypeConst {
    U8,
    U16,
    U32,
    U64,
    I8,
    I16,
    I32,
    I64,
    F32,
    F64,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Type {
    /// A type constant
    Const(Span, TypeConst),
    /// A type identifier: `T1`, `u16be`, `u32`
    Ident(Span, String),
    /// An array of the specified type, with a size: `[T; n]`
    Array(Span, Box<Type>, Expr),
    /// A union of types: `union { T, ... }`
    Union(Span, Vec<Type>),
    /// A struct type, with fields: `struct { field : T, ... }`
    Struct(Span, Vec<Field>),
}

impl Type {
    /// A type constant
    pub fn const_<Sp>(span: Sp, ty: TypeConst) -> Type
    where
        Sp: Into<Span>,
    {
        Type::Const(span.into(), ty)
    }

    /// A type identifier: `T1`, `u16be`, `u32`
    pub fn ident<Sp, S>(span: Sp, name: S) -> Type
    where
        Sp: Into<Span>,
        S: Into<String>,
    {
        Type::Ident(span.into(), name.into())
    }

    /// An array of the specified type, with a size: `[T; n]`
    pub fn array<Sp, T>(span: Sp, ty: T, size: Expr) -> Type
    where
        Sp: Into<Span>,
        T: Into<Box<Type>>,
    {
        Type::Array(span.into(), ty.into(), size)
    }

    /// A union of types: `union { T, ... }`
    pub fn union<Sp>(span: Sp, tys: Vec<Type>) -> Type
    where
        Sp: Into<Span>,
    {
        Type::Union(span.into(), tys)
    }

    /// A struct type, with fields: `struct { field : T, ... }`
    pub fn struct_<Sp>(span: Sp, fields: Vec<Field>) -> Type
    where
        Sp: Into<Span>,
    {
        Type::Struct(span.into(), fields)
    }
}

/// A field in a struct type
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Field {
    pub span: Span,
    pub name: String,
    pub ty: Type,
}

impl Field {
    pub fn new<Sp, S>(span: Sp, name: S, ty: Type) -> Field
    where
        Sp: Into<Span>,
        S: Into<String>,
    {
        let span = span.into();
        let name = name.into();

        Field { span, name, ty }
    }
}
