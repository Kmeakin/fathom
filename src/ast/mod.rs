//! The syntax of our data description language

use std::fmt;

pub mod binary;
pub mod host;

/// A variable that can either be free or bound
///
/// We use a locally nameless representation for variable binding.
///
/// # References
///
/// - [How I learned to stop worrying and love de Bruijn indices]
///   (http://disciple-devel.blogspot.com.au/2011/08/how-i-learned-to-stop-worrying-and-love.html)
/// - [The Locally Nameless Representation]
///   (https://www.chargueraud.org/research/2009/ln/main.pdf)
/// - [Locally nameless representation with cofinite quantification]
///   (http://www.chargueraud.org/softs/ln/)
/// - [A Locally-nameless Backend for Ott]
///   (http://www.di.ens.fr/~zappa/projects/ln_ott/)
/// - [Library STLC_Tutorial]
///   (https://www.cis.upenn.edu/~plclub/popl08-tutorial/code/coqdoc/STLC_Tutorial.html)
///
/// ## Libraries
///
/// There are a number of libraries out there for other languages that abstract
/// away handling locally nameless representations, but I've not yet figured out
/// how to port them to Rust yet:
///
/// - DBLib: Facilities for working with de Bruijn indices in Coq
///     - [Blog Post](http://gallium.inria.fr/blog/announcing-dblib/)
///     - [Github](https://github.com/coq-contribs/dblib)
/// - Bound: Bruijn indices for Haskell
///     - [Blog Post](https://www.schoolofhaskell.com/user/edwardk/bound)
///     - [Github](https://github.com/ekmett/bound/)
///     - [Hackage](https://hackage.haskell.org/package/bound)
/// - The Penn Locally Nameless Metatheory Library
///     - [Github](https://github.com/plclub/metalib)
#[derive(Clone, PartialEq, Eq)]
pub enum Var<N, B> {
    /// A free, unbound variable
    Free(N),
    /// A bound variable
    Bound(B),
}

impl<N, B> Var<N, B> {
    /// `true` if the variable is free
    fn is_closed(&self) -> bool {
        match *self {
            Var::Free(_) => true,
            Var::Bound(_) => false,
        }
    }

    pub fn abstract_with<F>(&mut self, f: &F)
    where
        F: Fn(&N) -> Option<B>,
    {
        *self = match *self {
            Var::Free(ref n) => match f(n) {
                None => return,
                Some(i) => Var::Bound(i),
            },
            Var::Bound(_) => return,
        }
    }
}

impl<F: fmt::Debug, B: fmt::Debug> fmt::Debug for Var<F, B> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match *self {
            Var::Free(ref x) => {
                write!(f, "Free(")?;
                x.fmt(f)?;
            }
            Var::Bound(ref i) => {
                write!(f, "Bound(")?;
                i.fmt(f)?;
            }
        }
        writeln!(f, ")")
    }
}

/// A variable with a name that is ignored for comparisons. This is useful for
/// improving error reporting when converting free varables to a named form.
///
/// # Type parameters
///
/// - `N`: The name that will be ignored for comparison purposes
/// - `T`: The type of the variable
#[derive(Clone, Eq, PartialOrd, Ord)]
pub struct Named<N, T>(pub N, pub T);

impl<N, T: PartialEq> PartialEq for Named<N, T> {
    fn eq(&self, other: &Named<N, T>) -> bool {
        self.1 == other.1
    }
}

impl<N, T: PartialEq> PartialEq<T> for Named<N, T> {
    fn eq(&self, other: &T) -> bool {
        &self.1 == other
    }
}

impl<N, T> From<(N, T)> for Named<N, T> {
    fn from(src: (N, T)) -> Named<N, T> {
        Named(src.0, src.1)
    }
}

impl<F: fmt::Debug, B: fmt::Debug> fmt::Debug for Named<F, B> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "Named(")?;
        self.0.fmt(f)?;
        write!(f, ", ")?;
        self.1.fmt(f)?;
        write!(f, ")")
    }
}

/// A field in a struct type
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Field<N, T> {
    pub name: N,
    pub value: T,
}

impl<N, T> Field<N, T> {
    pub fn new<M: Into<N>, U: Into<T>>(name: M, value: U) -> Field<N, T> {
        Field {
            name: name.into(),
            value: value.into(),
        }
    }

    /// Apply the function `f` to the field name and return the wrapped result
    pub fn map_name<M, F: FnMut(N) -> M>(self, mut f: F) -> Field<M, T> {
        Field {
            name: f(self.name),
            value: self.value,
        }
    }

    /// Apply the function `f` to the field value and return the wrapped result
    pub fn map_value<U, F: FnMut(T) -> U>(self, mut f: F) -> Field<N, U> {
        Field {
            name: self.name,
            value: f(self.value),
        }
    }
}
