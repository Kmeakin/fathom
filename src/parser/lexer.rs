use std::str::CharIndices;

use source::BytePos;
use unicode_xid::UnicodeXID;

fn is_ident_start(c: char) -> bool {
    UnicodeXID::is_xid_start(c) || c == '_'
}

fn is_ident_continue(c: char) -> bool {
    UnicodeXID::is_xid_continue(c) || c == '_'
}

fn is_digit(ch: char) -> bool {
    ch.is_digit(10)
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Error {
    pub location: BytePos,
    pub code: ErrorCode,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum ErrorCode {
    UnrecognizedToken,
}

fn error<T>(code: ErrorCode, location: BytePos) -> Result<T, Error> {
    Err(Error { location, code })
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum Token<'input> {
    // Data
    Ident(&'input str),
    IntLiteral(u64),

    // Symbols
    Equals, // =
    Semi, // ;
    Comma, // ,
    Pipe, // |
    Colon, // :
    Plus, // +
    Minus, // -
    Star, // *
    FSlash, // /

    // Delimeters
    LParen, // (
    RParen, // )
    LBrace, // {
    RBrace, // }
    LBracket, // [
    RBracket, // ]
}

pub struct Lexer<'input> {
    src: &'input str,
    chars: CharIndices<'input>,
    lookahead: Option<(usize, char)>,
}

impl<'input> Lexer<'input> {
    pub fn new(src: &'input str) -> Self {
        let mut chars = src.char_indices();

        Lexer {
            src,
            lookahead: chars.next(),
            chars,
        }
    }

    fn bump(&mut self) -> Option<(BytePos, char)> {
        let current = self.lookahead;
        self.lookahead = self.chars.next();
        current.map(|(index, ch)| (BytePos(index), ch))
    }

    fn slice(&self, start: BytePos, end: BytePos) -> &'input str {
        &self.src[start.0..end.0]
    }

    fn test_lookahead<F>(&self, mut pred: F) -> bool
    where
        F: FnMut(char) -> bool,
    {
        self.lookahead.map_or(false, |(_, ch)| pred(ch))
    }

    fn take_while<F>(&mut self, start: BytePos, mut keep_going: F) -> (BytePos, &'input str)
    where
        F: FnMut(char) -> bool,
    {
        self.take_until(start, |ch| !keep_going(ch))
    }

    fn take_until<F>(&mut self, start: BytePos, mut terminate: F) -> (BytePos, &'input str)
    where
        F: FnMut(char) -> bool,
    {
        while let Some((end, ch)) = self.lookahead {
            if terminate(ch) {
                return (BytePos(end), self.slice(start, BytePos(end)));
            } else {
                self.bump();
            }
        }

        let eof = BytePos(self.src.len());
        (eof, self.slice(start, eof))
    }

    fn ident(&mut self, start: BytePos) -> (BytePos, Token<'input>, BytePos) {
        let (end, ident) = self.take_while(start, is_ident_continue);
        (start, Token::Ident(ident), end)
    }

    fn int_literal(&mut self, start: BytePos) -> (BytePos, Token<'input>, BytePos) {
        let (end, int) = self.take_while(start, is_digit);
        (start, Token::IntLiteral(int.parse().unwrap()), end)
    }
}

impl<'input> Iterator for Lexer<'input> {
    type Item = Result<(BytePos, Token<'input>, BytePos), Error>;

    fn next(&mut self) -> Option<Result<(BytePos, Token<'input>, BytePos), Error>> {
        loop {
            return match self.bump() {
                Some((start, ch)) => {
                    let end = start.map(|x| x + 1);
                    match ch {
                        '=' => Some(Ok((start, Token::Equals, end))),
                        ';' => Some(Ok((start, Token::Semi, end))),
                        ',' => Some(Ok((start, Token::Comma, end))),
                        '|' => Some(Ok((start, Token::Pipe, end))),
                        ':' => Some(Ok((start, Token::Colon, end))),
                        '+' => Some(Ok((start, Token::Plus, end))),
                        '-' => Some(Ok((start, Token::Minus, end))),
                        '*' => Some(Ok((start, Token::Star, end))),
                        '/' if self.test_lookahead(|ch| ch == '/') => {
                            // Line comments
                            self.take_until(start, |ch| ch == '\n');
                            continue;
                        }
                        '/' => Some(Ok((start, Token::FSlash, end))),
                        '(' => Some(Ok((start, Token::LParen, end))),
                        ')' => Some(Ok((start, Token::RParen, end))),
                        '{' => Some(Ok((start, Token::LBrace, end))),
                        '}' => Some(Ok((start, Token::RBrace, end))),
                        '[' => Some(Ok((start, Token::LBracket, end))),
                        ']' => Some(Ok((start, Token::RBracket, end))),
                        ch if is_digit(ch) => Some(Ok(self.int_literal(start))),
                        ch if is_ident_start(ch) => Some(Ok(self.ident(start))),
                        ch if ch.is_whitespace() => continue,
                        _ => Some(error(ErrorCode::UnrecognizedToken, start)),
                    }
                }
                None => None,
            };
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    macro_rules! test {
        ($src:expr, $($span:expr => $token:expr,)*) => {{
            let lexed_tokens: Vec<_> = Lexer::new($src).collect();
            let expected_tokens = vec![$({
                let start = BytePos($span.find("~").unwrap());
                let end = BytePos($span.rfind("~").unwrap() + 1);
                Ok((start, $token, end))
            }),*];

            assert_eq!(lexed_tokens, expected_tokens);
        }};
    }

    #[test]
    fn ident() {
        test! {
            r#"  u8   "#,
            r#"  ~~   "# => Token::Ident("u8"),
        };
    }

    #[test]
    fn symbols() {
        test! {
            " = ; , | : + - * / ",
            " ~                 " => Token::Equals,
            "   ~               " => Token::Semi,
            "     ~             " => Token::Comma,
            "       ~           " => Token::Pipe,
            "         ~         " => Token::Colon,
            "           ~       " => Token::Plus,
            "             ~     " => Token::Minus,
            "               ~   " => Token::Star,
            "                 ~ " => Token::FSlash,
        }
    }

    #[test]
    fn delimiters() {
        test! {
            " ( ) { } [ ] ",
            " ~           " => Token::LParen,
            "   ~         " => Token::RParen,
            "     ~       " => Token::LBrace,
            "       ~     " => Token::RBrace,
            "         ~   " => Token::LBracket,
            "           ~ " => Token::RBracket,
        }
    }

    #[test]
    fn array_ty() {
        test! {
            "[u8; 34]",
            "~       " => Token::LBracket,
            " ~~     " => Token::Ident("u8"),
            "   ~    " => Token::Semi,
            "     ~~ " => Token::IntLiteral(34),
            "       ~" => Token::RBracket,
        };
    }

    #[test]
    fn struct_ty() {
        test! {
            "{ foo : u16 }",
            "~            " => Token::LBrace,
            "  ~~~        " => Token::Ident("foo"),
            "      ~      " => Token::Colon,
            "        ~~~  " => Token::Ident("u16"),
            "            ~" => Token::RBrace,
        };
    }
}
