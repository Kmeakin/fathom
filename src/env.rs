use std::collections::HashMap;

use ast::{Endianness, Type, TypeConst};

/// An environment of bindings and types
///
/// ```plain
/// Γ ::=
///         ·           empty environment
///         Γ, x:σ      environment extension
/// ```
#[derive(Debug)]
pub struct Env<'parent> {
    parent: Option<&'parent Env<'parent>>,
    tys: HashMap<String, Type>,
    bindings: HashMap<String, Type>,
}

impl<'parent> Env<'parent> {
    pub fn new() -> Env<'static> {
        Env {
            parent: None,
            tys: HashMap::new(),
            bindings: HashMap::new(),
        }
    }

    pub fn extend(&self) -> Env {
        Env {
            parent: Some(self),
            tys: HashMap::new(),
            bindings: HashMap::new(),
        }
    }

    pub fn add_ty<S>(&mut self, name: S, ty: Type)
    where
        S: Into<String>,
    {
        self.tys.insert(name.into(), ty);
    }

    pub fn add_binding<S>(&mut self, name: S, ty: Type)
    where
        S: Into<String>,
    {
        self.bindings.insert(name.into(), ty);
    }

    pub fn lookup_ty(&self, name: &str) -> Option<&Type> {
        self.tys.get(name).or_else(|| {
            self.parent.and_then(|env| env.lookup_ty(name))
        })
    }

    pub fn lookup_binding(&self, name: &str) -> Option<&Type> {
        self.bindings.get(name).or_else(|| {
            self.parent.and_then(|env| env.lookup_binding(name))
        })
    }
}

const BUILT_INS: [(&str, TypeConst); 30] = [
    ("u8", TypeConst::U(1, Endianness::Target)),
    ("u16", TypeConst::U(2, Endianness::Target)),
    ("u32", TypeConst::U(4, Endianness::Target)),
    ("u64", TypeConst::U(8, Endianness::Target)),
    ("i8", TypeConst::I(1, Endianness::Target)),
    ("i16", TypeConst::I(2, Endianness::Target)),
    ("i32", TypeConst::I(4, Endianness::Target)),
    ("i64", TypeConst::I(8, Endianness::Target)),
    ("f32", TypeConst::F(4, Endianness::Target)),
    ("f64", TypeConst::F(8, Endianness::Target)),

    ("u8le", TypeConst::U(1, Endianness::Little)),
    ("u16le", TypeConst::U(2, Endianness::Little)),
    ("u32le", TypeConst::U(4, Endianness::Little)),
    ("u64le", TypeConst::U(8, Endianness::Little)),
    ("i8le", TypeConst::I(1, Endianness::Little)),
    ("i16le", TypeConst::I(2, Endianness::Little)),
    ("i32le", TypeConst::I(4, Endianness::Little)),
    ("i64le", TypeConst::I(8, Endianness::Little)),
    ("f32le", TypeConst::F(4, Endianness::Little)),
    ("f64le", TypeConst::F(8, Endianness::Little)),

    ("u8be", TypeConst::U(1, Endianness::Big)),
    ("u16be", TypeConst::U(2, Endianness::Big)),
    ("u32be", TypeConst::U(4, Endianness::Big)),
    ("u64be", TypeConst::U(8, Endianness::Big)),
    ("i8be", TypeConst::I(1, Endianness::Big)),
    ("i16be", TypeConst::I(2, Endianness::Big)),
    ("i32be", TypeConst::I(4, Endianness::Big)),
    ("i64be", TypeConst::I(8, Endianness::Big)),
    ("f32be", TypeConst::F(4, Endianness::Big)),
    ("f64be", TypeConst::F(8, Endianness::Big)),
];

impl Default for Env<'static> {
    fn default() -> Env<'static> {
        let mut env = Env::new();

        for &(ident, ty_const) in &BUILT_INS {
            env.add_ty(ident, Type::Const(ty_const));
        }

        env
    }
}
