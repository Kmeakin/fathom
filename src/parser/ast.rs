//! The parse tree of the DDL
//!
//! This has a 1:1 mapping to the exposed concrete syntax. In order to do
//! anything interesting with this, one must first lower it to the core
//! representation.

use codespan::{BytePos, Span};

use parser::{self, grammar, ParseError};
use parser::lexer::Lexer;

pub use syntax::ast::{Binop, Const, FloatType, IntSuffix, TypeConst, Unop};

#[derive(Debug, Clone, PartialEq)]
pub struct Module<'input> {
    pub definitions: Vec<Definition<'input>>,
}

impl<'input> Module<'input> {
    /// Attempt to parse a module from a source string
    pub fn from_str(src: &'input str) -> Result<Module<'input>, ParseError> {
        grammar::parse_Module(Lexer::new(src).map(|x| x.map_err(ParseError::from)))
            .map_err(|err| parser::from_lalrpop(src, err))
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct Definition<'input> {
    pub doc: Vec<&'input str>,
    pub span: Span,
    pub name: &'input str,
    pub param_names: Vec<&'input str>,
    pub body_ty: Type<'input>,
}

impl<'input> Definition<'input> {
    /// Attempt to parse a definition from a source string
    pub fn from_str(src: &'input str) -> Result<Definition<'input>, ParseError> {
        grammar::parse_Definition(Lexer::new(src).map(|x| x.map_err(ParseError::from)))
            .map_err(|err| parser::from_lalrpop(src, err))
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct Field<'input, T> {
    pub doc: Vec<&'input str>,
    pub name: &'input str,
    pub value: T,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Type<'input> {
    Var(Span, &'input str),
    Array(Span, Box<Type<'input>>, Box<Expr<'input>>),
    Cond(Span, Vec<Field<'input, (Expr<'input>, Type<'input>)>>),
    Struct(Span, Vec<Field<'input, Type<'input>>>),
    Where(
        Span,
        Box<Type<'input>>,
        BytePos,
        &'input str,
        Box<Expr<'input>>,
    ),
    Compute(Span, TypeConst, Box<Expr<'input>>),
    App(Span, Box<Type<'input>>, Vec<Type<'input>>),
}

impl<'input> Type<'input> {
    /// Attempt to parse a type from a source string
    pub fn from_str(src: &'input str) -> Result<Type<'input>, ParseError> {
        grammar::parse_Type(Lexer::new(src).map(|x| x.map_err(ParseError::from)))
            .map_err(|err| parser::from_lalrpop(src, err))
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum Expr<'input> {
    Ann(Span, Box<Expr<'input>>, TypeConst),
    Const(Span, Const),
    Var(Span, &'input str),
    Unop(Span, Unop, Box<Expr<'input>>),
    Binop(Span, Binop, Box<Expr<'input>>, Box<Expr<'input>>),
    Array(Span, Vec<Expr<'input>>),
    Proj(Span, Box<Expr<'input>>, &'input str),
    Subscript(Span, Box<Expr<'input>>, Box<Expr<'input>>),
    Cast(Span, Box<Expr<'input>>, TypeConst),
}

impl<'input> Expr<'input> {
    /// Attempt to parse an expression from a source string
    pub fn from_str(src: &'input str) -> Result<Expr<'input>, ParseError> {
        grammar::parse_Expr(Lexer::new(src).map(|x| x.map_err(ParseError::from)))
            .map_err(|err| parser::from_lalrpop(src, err))
    }
}
