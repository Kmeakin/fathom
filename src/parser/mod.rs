use lalrpop_util;

use ast::{BoolExpr, Definition, Expr, Type};
use env::Env;
use source::BytePos;

mod lexer;
#[allow(unused_extern_crates)]
mod grammar;

use self::lexer::{Lexer, Error as LexerError, Token};

pub type ParseError<'input> = lalrpop_util::ParseError<BytePos, Token<'input>, LexerError>;

// pub enum ParseError<L, T, E> {
//     InvalidToken {
//         location: L,
//     },
//     UnrecognizedToken {
//         token: Option<(L, T, L)>,
//         expected: Vec<String>,
//     },
//     ExtraToken {
//         token: (L, T, L),
//     },
//     User {
//         error: E,
//     },
// }

pub fn parse<'input, 'env>(
    env: &'env Env,
    src: &'input str,
) -> Result<Vec<Definition>, ParseError<'input>> {
    grammar::parse_Definitions(env, Lexer::new(src))
}

pub fn parse_expr<'input, 'env>(
    env: &'env Env,
    src: &'input str,
) -> Result<Expr, ParseError<'input>> {
    grammar::parse_Expr(env, Lexer::new(src))
}

pub fn parse_bool_expr<'input, 'env>(
    env: &'env Env,
    src: &'input str,
) -> Result<BoolExpr, ParseError<'input>> {
    grammar::parse_BoolExpr(env, Lexer::new(src))
}

pub fn parse_ty<'input, 'env>(
    env: &'env Env,
    src: &'input str,
) -> Result<Type, ParseError<'input>> {
    grammar::parse_Type(env, Lexer::new(src))
}

#[cfg(test)]
mod tests {
    use env::Env;
    use super::*;

    #[test]
    fn parse_bool_expr_atomic() {
        let src = "
            !((true | (false)))
        ";

        assert_snapshot!(parse_bool_expr, parse_bool_expr(&Env::default(), src));
    }

    #[test]
    fn parse_bool_expr_operators() {
        let src = "
            (true & false) | (true | false)
        ";

        assert_snapshot!(
            parse_bool_expr_operators,
            parse_bool_expr(&Env::default(), src)
        );
    }

    #[test]
    fn parse_add_expr() {
        let src = "x + y + z";

        assert_snapshot!(parse_add_expr, parse_expr(&Env::default(), src));
    }

    #[test]
    fn parse_sub_expr() {
        let src = "x - y - z";

        assert_snapshot!(parse_sub_expr, parse_expr(&Env::default(), src));
    }

    #[test]
    fn parse_add_expr_mixed() {
        let src = "x + y + z - z + x";

        assert_snapshot!(parse_add_expr_mixed, parse_expr(&Env::default(), src));
    }

    #[test]
    fn parse_mul_expr() {
        let src = "x * y * z";

        assert_snapshot!(parse_mul_expr, parse_expr(&Env::default(), src));
    }

    #[test]
    fn parse_div_expr() {
        let src = "x / y / z";

        assert_snapshot!(parse_div_expr, parse_expr(&Env::default(), src));
    }

    #[test]
    fn parse_mul_expr_mixed() {
        let src = "x * y * z / z * x";

        assert_snapshot!(parse_mul_expr_mixed, parse_expr(&Env::default(), src));
    }

    #[test]
    fn parse_mixed_arithmetic_expr() {
        let src = "x + y * z / z - x * a";

        assert_snapshot!(
            parse_mixed_arithmetic_expr,
            parse_expr(&Env::default(), src)
        );
    }

    #[test]
    fn parse_ty_var() {
        let src = "
            Point
        ";

        assert_snapshot!(parse_ty_var, parse_ty(&Env::default(), src));
    }

    #[test]
    fn parse_ty_empty_struct() {
        let src = "struct {}";

        assert_snapshot!(parse_ty_empty_struct, parse_ty(&Env::default(), src));
    }

    #[test]
    fn parse_ty_where() {
        let src = "
            struct {
                x: u32 where x => true,
            }
            where x => true
            where x => false
        ";

        assert_snapshot!(parse_ty_where, parse_ty(&Env::default(), src));
    }

    #[test]
    fn parse_simple_definition() {
        let src = "
            Offset32 = u32;
        ";

        assert_snapshot!(parse_simple_definition, parse(&Env::default(), src));
    }

    #[test]
    fn parse_array_with_constant_size() {
        let src = "
            Point = [f32; 3];
        ";

        assert_snapshot!(parse_array_with_constant_size, parse(&Env::default(), src));
    }

    #[test]
    fn parse_definition() {
        let src = "
            Point = struct {
                x : u32be,
                y : u32be,
            };

            Array = struct {
                len : u16le,
                data : [Point; len],
            };

            Formats = union {
                struct { format : u16, data: u16 },
                struct { format : u16, point: Point },
                struct { format : u16, array: Array },
            };
        ";

        assert_snapshot!(parse_definition, parse(&Env::default(), src));
    }
}
