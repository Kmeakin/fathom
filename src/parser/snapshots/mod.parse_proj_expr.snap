Unop(
    Neg,
    Proj(
        Proj(
            Var(
                Free("foo")
            ),
            "bar"
        ),
        "x"
    )
)
